### PR TITLE
Fixes #2617, fixes addr of obj downcast.

### LINKED ISSUE #2617: [JS][Codegen] Initialising object doesn't create unmentioned fields.
```nim
type T = object
    a, b: int

let v = T(a: 123)
echo v.b
```
```sh
$ nim js -d:nodejs test.nim
$ node nimcache/test.js 
undefined
```

### PR DESCRIPTION
**Pull request by yglukhov**
Two fixes here: one for #2617. One for genAddr of object downcast.

### CODE REVIEW

File: compiler/jsgen.nim [hunk 1]
```diff
@@ -1096,24 +1098,34 @@ proc putToSeq(s: string, indirect: bool): Rope =
...
-      add(result, createRecordVarAux(p, lastSon(rec.sons[i]), c))
+      createRecordVarAux(p, lastSon(rec.sons[i]), excludeMembers, output)
   of nkSym:
-    if c > 0: add(result, ", ")
-    add(result, mangleName(rec.sym))
-    add(result, ": ")
-    add(result, createVar(p, rec.sym.typ, false))
-    inc(c)
+    let name = $mangleName(rec.sym)
+    if not (name in excludeMembers):
```
Comment (Araq) [inline compiler/jsgen.nim]: if name notin excludeMembers

File: compiler/jsgen.nim [hunk 2]
```diff
@@ -33,7 +33,7 @@ import
   ast, astalgo, strutils, hashes, trees, platform, magicsys, extccomp, options,
   nversion, nimsets, msgs, securehash, bitsets, idents, lists, types, os,
   times, ropes, math, passes, ccgutils, wordrecg, renderer, rodread, rodutils,
-  intsets, cgmeth, lowerings
+  intsets, cgmeth, lowerings, sets
```
Comment (Araq) [inline compiler/jsgen.nim]: Just use an intset instead please.

File: compiler/jsgen.nim [hunk 3]
```diff
@@ -1435,6 +1442,7 @@ proc genObjConstr(p: PProc, n: PNode, r: var TCompRes) =
   var a: TCompRes
   r.res = rope("{")
   r.kind = resExpr
+  var fieldNames = initSet[string]()
```
Comment (Araq) [inline compiler/jsgen.nim]: Use symbol IDs instead of symbol names.

**Comment by yglukhov** [comment]
Good point. Done.

**Comment by Araq** [comment]
but why do you need an exclude set in the first place?

**Comment by yglukhov** [comment]
```nim
var a = MyObj(myField: 0)
```
is generated by ``genObjConstr``, which previously emitted only those fields listed in node. Now it also emits fields from ``createObjInitList``, telling ``createObjInitList`` not to emit fields it has already emitted itself.

**Comment by yglukhov** [comment]
Thank you.

**Pull request merged**